src/compile_target.zig

```diff
@@ -0,0 +1,439 @@
+                    _ = libarchive.Archive.extractToDir(
+                        compressed_archive_bytes.list.items,
+                        tmpdir,
+                        null,
+                        void,
+                        {},
+                        // "package/bin"
+                        2,
+                        true,
+                        false,
+                    ) catch |err| {
+                        node.end();
+                        Output.err(err,
+                            \\Failed to extract the downloaded tarball
+                            \\
+                            \\After downloading {} from {s}
+                        , .{
+                            this.*,
+                            url_str,
+                        });
+                        Global.exit(1);
+                    };
+
+                    var did_retry = false;
+                    while (true) {
+                        bun.C.moveFileZ(bun.toFD(tmpdir), if (this.os == .windows) "bun.exe" else "bun", bun.invalid_fd, dest_z) catch |err| {
+                            if (!did_retry) {
+                                did_retry = true;
+                                const dirname = bun.path.dirname(dest_z, .loose);
+                                if (dirname.len > 0) {
+                                    std.fs.cwd().makePath(dirname) catch {};
+                                }
+                                continue;
+                            }
+                            node.end();
+                            Output.err(err, "Failed to move cross-compiled bun binary into cache directory {}", .{bun.fmt.fmtPath(u8, dest_z, .{})});
+                            Global.exit(1);
+                        };
+                        break;
+                    }
+                }
+                refresher.refresh();
+            }
+        }
+    }
+
+    pub fn isSupported(this: *const CompileTarget) bool {
+        return switch (this.os) {
+            .windows => this.arch == .x64,
+            .mac => true,
+            .linux => this.libc == .default,
+            .wasm => false,
+        };
+    }
+
+    pub fn from(input_: []const u8) CompileTarget {
+        var this = CompileTarget{};
+
+        const input = bun.strings.trim(input_, " \t\r");
+        if (input.len == 0) {
+            return this;
+        }
+
+        var found_os = false;
+        var found_arch = false;
+        var found_baseline = false;
+        var found_version = false;
+        var found_libc = false;
+
+        // Parse each of the supported values.
+        // The user shouldn't have to care about the order of the values. As long as it starts with "bun-".
+        // Nobody wants to remember whether its "bun-linux-x64" or "bun-x64-linux".
+        var splitter = bun.strings.split(input, "-");
+        while ((!found_arch or !found_os or !found_baseline or !found_version or !found_libc) and input.len > 0) {
```

Comment:
this parser is way too loose.

the following inputs behave differently than i would have expected prior to reading this code.

- `bun-dsfkjfdsajd-x64----windows-wasm-aarch64-32bit-mac` downloads `bun-windows-x64`
- `bun-x86` parses as aarch64
- `bun--` parses

Reply:
`bun-` and keyboard-mashing will no longer pass

but `bun--` seems harmless to ignore